ReactAndroid/src/main/java/com/facebook/react/views/webview/ReactWebViewManager.java

```diff
@@ -133,19 +134,61 @@ public void onPageStarted(WebView webView, String url, Bitmap favicon) {
 
     @Override
     public boolean shouldOverrideUrlLoading(WebView view, String url) {
-        if (url.startsWith("http://") || url.startsWith("https://") ||
-            url.startsWith("file://")) {
-          return false;
-        } else {
-          try {
-            Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
-            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
-            view.getContext().startActivity(intent);
-          } catch (ActivityNotFoundException e) {
-            FLog.w(ReactConstants.TAG, "activity not found to handle uri scheme for: " + url, e);
+      ReadableMap[] policies = ((ReactWebView) view).getNavigationBlockingPolicies();
+      ReadableMap policy;
+      boolean policyFulfilled = false;
+
+      if (policies != null) {
+        WritableMap currentValues = createWebViewEvent(view, url);
+        currentValues.putString("currentURL", view.getUrl());
+
+        for (int i = 0; i < policies.length; i++) {
+          policy = policies[i];
+          ReadableMapKeySetIterator it = policy.keySetIterator();
+          String policyKey;
+
+          // has at least one rule
+          if (it != null && it.hasNextKey()) {
+            policyFulfilled = true;
+
+            // loop until there are no more rules or one has been rejected already
+            while (it.hasNextKey() && policyFulfilled) {
+              policyKey = it.nextKey();
+
+              if (!currentValues.hasKey(policyKey)) {
+                policyFulfilled = false;
+                continue;
+              }
+
+              policyFulfilled = currentValues.getString(policyKey)
+                .matches(policy.getString(policyKey));
+            }
+          }
+          
+          if (policyFulfilled) {
+            dispatchEvent(
+              view,
+              new TopNavigationBlockedEvent(
+                view.getId(),
+                createWebViewEvent(view, url)));
+            return true;
           }
-          return true;
         }
+      }
+
+      if (url.startsWith("http://") || url.startsWith("https://") ||
+          url.startsWith("file://")) {
+        return false;
+      } else {
+        try {
+          Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
+          intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+          view.getContext().startActivity(intent);
+        } catch (ActivityNotFoundException e) {
+          FLog.w(ReactConstants.TAG, "activity not found to handle uri scheme for: " + url, e);
+        }
+        return true;
+      }
```

Comment:
Same explanation as before. If we think this is the way to go, sure, it makes sense.